packages/dds/merge-tree/Obliterate.md

```diff
@@ -0,0 +1,276 @@
+```
+{ seq: 60, refSeq: 40, clientId: 3, op: <insert "hello" at index 10> }
+```
+
+After locating the insertion point and updating the merge tree, we need to decide if the resulting segment is inside of an obliterated region.
+If we happened to know the `seq` of the obliteration we were testing for, this is easy: the first adjacent segment in each direction from
+the perspective of `{ seq: 50, clientId: localClientId }` can inform us if we're either inside or directly adjacent to an obliterated range.
+Thus, a naive implementation could check all sequence numbers in the collab window.
+The obvious optimization of only checking seq numbers of obliterate ops would improve this slightly.
+We can do asymptotically better by leveraging the tree structure.
+It would be ideal if we only needed to perform one commonly short excursion in each direction.
+The only candidate that makes much sense is from the perspective of `{ seq: 60, clientId: 3 }` (i.e. the client submitting the insert op at
+the time the op is sequenced).
+The problem with this perspective is that ops 51 through 59 may have inserted a segment between the inserted "hello"
+and the obliterated range that was submitted by a client which has already acked the obliterate.
+For example, `{ seq: 55, clientId: 5, refSeq: 50, op: <insert "i won't be obliterated" at index 10> }`.
+
+The forward excursion would need to continue past this segment in order to conclude it isn't in an obliterated range.
+If that was the only such concurrent insert, the next segment it would visit would be an obliterated one and we'd decide
+whether or not to include the newly inserted segment as part of the obliterated region based on some endpoint merge strategy.
+
+The key insight is that visiting the segment with seq 55 does provide the excursion with information: since the segment
+was inserted at seq 55 and isn't obliterated, any obliterate operation must have occurred before seq 55.
+If we keep track of the smallest sequence number that we've visited, we can halt the excursion as soon as it falls below
+the smallest obliterate operation within the collab window.
+If we alternatively reach a segment that has been obliterated concurrently to the insert we're processing, we can also stop
+and use the endpoint resolution strategy.
+
+All-in-all, the insert logic modification might look something like this:
+
+```typescript
+function insertingWalk(args /* mostly omitted */, op) {
+  /* regular insert logic goes here */
+
+  let currentMin = Number.POSITIVE_INFINITY;
+  let obliteratedSegment: ISegment | undefined = undefined;
+  const smallestSeqObliterateOp = this.getSmallestSeqObliterateOp();
+  const findAdjacentObliteratedSegment = (seg) => {
+    if (seg.seq === UnassignedSequenceNumber) {
+      // Ignore un-acked segments
+      return true;
+    }
+
+    if (seg.obliteratedSeq && seg.obliteratedSeq > op.referenceSequenceNumber) {
```

Comment:
what will seg.obliteratedSeq be in the case of overlapping obliterates on a segment?

Reply:
I think you're right in the other comment that it would need to be the `max`, but actually for move semantics I'm not sure one number is sufficient. First thought is that we need to record all of them in the collab window, since the segment should probably move with the first move after the insert's refSeq.

Reply:
actually, for move one number should be fine, and I think we do want the smaller of the two: if the segments on either side were both concurrently moved to different locations, we should move this segment with the one that happened first.

That said, if we don't implement customizable endpoint semantics and we opt for half-open expansion of the moved range, we should only ever need to look in one direction anyway.

Reply:
This is spec'd out more clearly now with a worked through example of how overlapping obliterate might look in partial lengths bookkeeping. `movedSeq`'s behavior matches `removedSeq` in that the first mover wins, and other overlapping movers/obliteraters are stored in `movedClientIds` (like `removedClientIds`)